# Изучение набора данных


> 🚧 Раздел в разработке.


Перед анализом важно “познакомиться” с данными: понять, **что хранится в таблице**, **какая геометрия**, **в какой системе координат**, **есть ли пропуски и ошибки**, и **какие атрибуты доступны для расчётов**.

На примере одного из наборов пространственных данных в формате **GeoDataFrame**, мы посмотрим, как это можно сделать


## 0. Импорт библиотек


In [ ]:
import geopandas as gpd

- [**GeoPandas**](https://geopandas.org/) (`geopandas`) — расширение библиотеки pandas, предназначенное для работы с геопространственными данными. Позволяет загружать, обрабатывать и анализировать пространственные наборы данных в различных форматах.


## 1. Основная информация


Прочитаем данные из формата `.geojson`


In [13]:
gdf_geojson = gpd.read_file("data/spb_metro.geojson") 

### 1.1 `head()`

Метод `head()` отображает первые строки `DataFrame` или `GeoDataFrame`.
Он используется для первичного знакомства с содержимым данных.

По умолчанию выводятся первые **5 строк**.


In [14]:
gdf_geojson.head()

,field_1,number,name,address,note,lat,lon,geometry
0,1,1,пр.Ветеранов-1,"г. Санкт-Петербург, бульвар Новаторов, д.98 в",-,59.841346,30.253193,POINT (30.25319 59.84135)
1,2,2,пр.Ветеранов-2,"г. Санкт-Петербург, бульвар Новаторов, д.108 а",-,59.841500,30.250561,POINT (30.25056 59.8415)
2,3,3,Ленинский пр.-1,"г. Санкт-Петербург, Ленинский пр., д.127",-,59.851345,30.266748,POINT (30.26675 59.85134)
3,4,4,Ленинский пр.-2,"г. Санкт-Петербург, бульвар Новаторов, д.10",-,59.849560,30.269506,POINT (30.26951 59.84956)
4,5,5,Автово,"г. Санкт-Петербург, пр.Стачек, д.90 корпус 2",-,59.867313,30.261358,POINT (30.26136 59.86731)


Можно указать количество строк вручную:


In [15]:
gdf_geojson.head(10)

,field_1,number,name,address,note,lat,lon,geometry
0,1,1,пр.Ветеранов-1,"г. Санкт-Петербург, бульвар Новаторов, д.98 в",-,59.841346,30.253193,POINT (30.25319 59.84135)
1,2,2,пр.Ветеранов-2,"г. Санкт-Петербург, бульвар Новаторов, д.108 а",-,59.841500,30.250561,POINT (30.25056 59.8415)
2,3,3,Ленинский пр.-1,"г. Санкт-Петербург, Ленинский пр., д.127",-,59.851345,30.266748,POINT (30.26675 59.85134)
3,4,4,Ленинский пр.-2,"г. Санкт-Петербург, бульвар Новаторов, д.10",-,59.849560,30.269506,POINT (30.26951 59.84956)
4,5,5,Автово,"г. Санкт-Петербург, пр.Стачек, д.90 корпус 2",-,59.867313,30.261358,POINT (30.26136 59.86731)
5,6,6,Кировский з-д,"г. Санкт-Петербург, пр.Стачек, д.70",-,59.879602,30.262023,POINT (30.26202 59.8796)
6,7,7,Нарвская,"г. Санкт-Петербург, пл. Стачек, д.2",-,59.901170,30.274932,POINT (30.27493 59.90117)
7,8,8,Балтийская,"г. Санкт-Петербург, пл.Балтийского вокзала, д.1",-,59.907212,30.299465,POINT (30.29946 59.90721)
8,9,9,Пушкинская,"г. Санкт-Петербург, Загородный проспект, 52А",-,59.920542,30.329747,POINT (30.32975 59.92054)
9,10,10,пл.Восстания-1,"г. Санкт-Петербург, пл.Восстания, д.2",-,59.931580,30.360532,POINT (30.36053 59.93158)


### 1.2 `info()`

Метод `info()` отображает краткую сводную информацию о `DataFrame` или `GeoDataFrame`:

- количество записей (строк),
- названия столбцов,
- типы данных в каждом столбце,
- количество **непустых значений**,
- объём памяти, занимаемый данными.


In [16]:
gdf_geojson.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 75 entries, 0 to 74
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   field_1   75 non-null     int32   
 1   number    75 non-null     int32   
 2   name      75 non-null     object  
 3   address   75 non-null     object  
 4   note      75 non-null     object  
 5   lat       75 non-null     float64 
 6   lon       75 non-null     float64 
 7   geometry  75 non-null     geometry
dtypes: float64(2), geometry(1), int32(2), object(3)
memory usage: 4.2+ KB


### 1.3 `len()` и `shape`

Чтобы понять размер набора данных, удобно использовать метод `len()` и атрибут `shape`.

Функция `len()` возвращает количество строк в `DataFrame` или `GeoDataFrame`,
то есть **количество пространственных объектов**.


In [17]:
len(gdf_geojson)

75

Атрибут `shape` возвращает кортеж из двух чисел:

- первое число — количество строк (объектов),
- второе число — количество столбцов (атрибутов, включая `geometry`).


In [18]:
gdf_geojson.shape

(75, 8)

## 2. Геометрия


В `GeoDataFrame` каждый объект имеет **геометрию**

Перед выполнением пространственного анализа важно ответить на несколько вопросов:

- Какие типы геометрии содержатся в данных (точки, линии, полигоны)?
- Корректны ли геометрии?


### 2.2 `geometry`

В `GeoDataFrame` пространственная информация хранится в специальном столбце `geometry`.
Именно он отличает `GeoDataFrame` от обычного `DataFrame`.


Этот атрибут возвращает объект типа `GeoSeries`, содержащий геометрию каждого пространственного объекта.


In [19]:
gdf_geojson.geometry

0     POINT (30.25319 59.84135)
1      POINT (30.25056 59.8415)
2     POINT (30.26675 59.85134)
3     POINT (30.26951 59.84956)
4     POINT (30.26136 59.86731)
                ...            
70    POINT (30.29028 59.95172)
71      POINT (30.292 59.96095)
72    POINT (30.25955 59.97182)
73    POINT (30.25517 59.98944)
74    POINT (30.25983 60.00906)
Name: geometry, Length: 75, dtype: geometry

Иногда в наборе данных встречаются объекты без геометрии.
Это критично, потому что такие строки не могут участвовать в пространственном анализе.

При работе с пространственными данными важно различать:

- `NaN` — геометрия отсутствует,
- `empty geometry` — геометрия формально существует, но она пустая.


#### 2.1.1 Проверка пропусков (`NaN`)

Метод:

- `isna()` — определяет пропущенные значения,
- `sum()` — подсчитывает их количество.


In [20]:
gdf_geojson.geometry.isna().sum()

0

Если значение больше нуля — часть строк содержит `NaN`, и их необходимо обработать (например, удалить или восстановить геометрию).

Удалить можно, например, так:


In [21]:
gdf_geojson = gdf_geojson.dropna(subset=["geometry"])

#### 2.1.2 Проверка пустых геометрий


In [22]:
gdf_geojson.geometry.is_empty.sum()

0

#### 2.1.3 Поле, где записана геометрия

В `GeoDataFrame` всегда есть активное поле геометрии — именно он используется для всех пространственных операций.

По умолчанию его имя — `geometry`, но оно может быть другим (например, после переименования или создания новой геометрии).


In [ ]:
gdf_geojson.geometry.name

- В `GeoDataFrame` может быть несколько столбцов с геометриями, но только один из них является активным.
- Все пространственные операции (`buffer`, `intersection`, `area`, `plot`) выполняются именно над активным столбцом геометрии.


### 2.2. `geom_type`


Чтобы определить тип геометрии объектов, используется атрибут `geom_type`. Он возвращает серию (`GeoSeries`), в которой для каждой строки указан тип геометрии:


gdf_geojson.geom_type.unique()


Посмотрим, какие уникальные типы встречаются в данных


In [23]:
gdf_geojson.geom_type.value_counts()

Point    75
Name: count, dtype: int64

### 2.2. `is_valid`


Помимо пропусков и пустых объектов, геометрия может быть **невалидной** (геометрически некорректной).

Чаще всего проблемы с некорректной геометрией возникают у полигонов, например:

- самопересечение (self-intersection),
- незамкнутые контуры,
- дублирующиеся вершины,
- некорректная топология.


Можно проверить количество некорректных геометрий в наборе данных:

- метод `.is_valid` — возвращает `True` или `False` для каждой геометрии;
- `~` — логическое отрицание (находит невалидные объекты);
- `sum()` — подсчитывает их количество.


In [24]:
(~gdf_geojson.is_valid).sum()

0

Посмотреть, какие именно объекты имеют некорректную геометрию


In [25]:
gdf_geojson[~gdf_geojson.is_valid]

,field_1,number,name,address,note,lat,lon,geometry


Исправить некорректную геометрию можно так


In [26]:
gdf_geojson["geometry"] = gdf_geojson.make_valid()

После исправления стоит повторно проверить


In [27]:
(~gdf_geojson.is_valid).sum()

0

## 3. Система координат (CRS)


Координаты для геометрий в `GeoDataFrame` записаны в определённой **системе координат (CRS — Coordinate Reference System)**.

Перед выполнением пространственного анализа важно ответить на несколько вопросов:

- Указана ли система координат в данных?
- Какой у неё тип?
- Подходит ли текущая CRS для расчёта расстояний и площадей?
- Совпадает ли CRS у всех используемых слоёв?

Здесь мы рассмотрим основные характеристики CRS, которые можно определить на этапе первичного анализа. Подробнее о системах координат поговорим во втором модуле.


### 3.1 `crs`


Атрибут `crs` позволяет узнать, в какой системе координат записаны пространственные данные.


In [28]:
gdf_geojson.crs

<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World.
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

CRS (Coordinate Reference System) определяет:

- систему отсчёта координат,
- единицы измерения (градусы или метры),
- способ проецирования поверхности Земли на плоскость.


### 3.2 Тип системы координат


Система координат может быть:

- **географической** — координаты заданы в градусах (широта и долгота),
- **проекционной** — координаты заданы в линейных единицах (чаще всего в метрах).

Проверить тип CRS можно так:


In [31]:
print("Географическая:", gdf_geojson.crs.is_geographic)
print("Проекционная:", gdf_geojson.crs.is_projected)

Географическая: True
Проекционная: False


- Если `is_geographic == True` — координаты в градусах (например, `EPSG:4326`).
- Если `is_projected == True` — координаты в метрах или других линейных единицах.


### 3.3 Единицы измерения координат


Можно узнать, в каких единицах записаны координаты:

- `degree` — градусы,
- `metre` — метры
  могут быть указаны и другие единицы измерения


In [33]:
gdf_geojson.crs.axis_info

[Axis(name=Geodetic latitude, abbrev=Lat, direction=north, unit_auth_code=EPSG, unit_code=9122, unit_name=degree),
 Axis(name=Geodetic longitude, abbrev=Lon, direction=east, unit_auth_code=EPSG, unit_code=9122, unit_name=degree)]

- `degree` — градусы,
- `metre` — метры


### 3.4 Получение EPSG-кода


Иногда удобно получить только числовой код системы координат:


In [34]:
gdf_geojson.crs.to_epsg()

4326

Если CRS нестандартная, метод может вернуть `None`.


## 4. Атрибутивная информация


Помимо геометрии, каждый объект в `GeoDataFrame` содержит **атрибутивную информацию** — описательные характеристики, связанные с пространственными объектами.

В этом разделе мы рассмотрим, как изучать структуру атрибутивных данных, проверять типы полей, искать пропуски и получать базовую статистику.


### 4.1 `columns`

Атрибут `columns` позволяет получить список всех столбцов в `DataFrame` или `GeoDataFrame`.
Результат представляет собой объект типа `Index`, содержащий названия всех полей, включая поле с геометрией:


In [35]:
gdf_geojson.columns

Index(['field_1', 'number', 'name', 'address', 'note', 'lat', 'lon',
       'geometry'],
      dtype='object')

### 4.2 `dtypes`

Атрибут `dtypes` позволяет узнать тип данных каждого столбца в `DataFrame` или `GeoDataFrame`.


In [36]:
gdf_geojson.dtypes

field_1        int32
number         int32
name          object
address       object
note          object
lat          float64
lon          float64
geometry    geometry
dtype: object

Типы данных необходимо проверять, чтобы убедиться, что каждое поле корректный формат и с ним можно выполнять соответствующие аналитические операции без ошибок


## 5. Другие методы изучения

Рассмотрим методы, которые не относятся к категориям выше.


### 5.1 Границы набора данных

**Bounding Box** — это минимальный прямоугольник, который полностью охватывает все объекты набора данных.

Он задаётся четырьмя координатами:

```
[minx, miny, maxx, maxy]
```

Получить границы всего набора данных можно так:


In [37]:
gdf_geojson.total_bounds

array([30.202357, 59.815405, 30.500607, 60.068754])

## 6. Итог


В этом разделе мы познакомились с основными способами первичного анализа пространственного набора данных в формате `GeoDataFrame`.

Это очень важный этап проверк загруженных данных, так как позволяет понять структуру и содержание данных, убедиться в корректности загрузки, выявить потенциальные ошибки
